### PR TITLE
Laboratory work #3, Sofya Kostina - 21FPL2

### PR DESCRIPTION
**Pull request by d07l**

### CODE REVIEW

File: lab_3_keywords_textrank/main.py [hunk 1]
```diff
@@ -51,7 +52,12 @@ def _clean_and_tokenize(self, text: str) -> tuple[str, ...]:
             tuple[str, ...]
                 clean lowercase tokens
         """
-        pass
+        for i in text:
+            if i in self._punctuation:
+                text = text.replace(i, ' ')
+        splited_text = text.split()
+        tokens = [token.lower() for token in splited_text]
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:59]: It's better to call `.lower()` for all `text` before `split`

File: lab_3_keywords_textrank/main.py [hunk 2]
```diff
@@ -136,7 +148,16 @@ def encode(self, tokens: tuple[str, ...]) -> Optional[tuple[int, ...]]:
...
         """
-        pass
+        if not tokens:
+            return None
+
+        self._learn_indices(tokens)
+        numbers_tokens = []
+        for token in tokens:
+            if token in self._word2id:
+                numbers_tokens.append(self._word2id.get(token))
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:158]: You don't need to use the `get` method if you check that the `token` is in the `self._word2id`

File: lab_3_keywords_textrank/main.py [hunk 3]
```diff
@@ -152,8 +173,14 @@ def decode(self, encoded_tokens: tuple[int, ...]) -> Optional[tuple[str, ...]]:
                 sequence of string tokens
         In case of out-of-dictionary input data, None is returned
         """
-        pass
+        decoded_tokens = []
+        for token in encoded_tokens:
+            if token in self._id2word:
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:178]: revert the condition and you will not need to use `else`

File: lab_3_keywords_textrank/main.py [hunk 4]
```diff
@@ -173,7 +200,32 @@ def extract_pairs(tokens: tuple[int, ...], window_length: int) -> Optional[tuple
...
+
+    for count, value in enumerate(tokens):
+        window = tokens[count:window_length + count]
+        windows.append(window)
+
+    for window in windows:
+        if len(window) != window_length:
+            continue
+
+        for i in range(0, len(window) - 1):
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:218]: 0 is not needed

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_keywords_textrank/main.py (hunk 1 above)
Comment (d07l) [reply lab_3_keywords_textrank/main.py]: fixed

File: lab_3_keywords_textrank/main.py (hunk 2 above)
Comment (d07l) [reply lab_3_keywords_textrank/main.py]: fixed

File: lab_3_keywords_textrank/main.py (hunk 3 above)
Comment (d07l) [reply lab_3_keywords_textrank/main.py]: fixed

File: lab_3_keywords_textrank/main.py (hunk 4 above)
Comment (d07l) [reply lab_3_keywords_textrank/main.py]: fixed

File: lab_3_keywords_textrank/main.py [hunk 5]
```diff
@@ -51,7 +52,12 @@ def _clean_and_tokenize(self, text: str) -> tuple[str, ...]:
             tuple[str, ...]
                 clean lowercase tokens
         """
-        pass
+        for i in text:
+            if i in self._punctuation:
+                text = text.replace(i, ' ')
+        splited_text = text.lower().split()
+        tokens = [token for token in splited_text]
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:59]: You don't need to generate list before tuple. Let's generate tuple immediately

Comment (artyomtugaryov) [reply lab_3_keywords_textrank/main.py:59]: Like in `_remove_stop_words`

File: lab_3_keywords_textrank/main.py [hunk 6]
```diff
@@ -136,7 +148,16 @@ def encode(self, tokens: tuple[str, ...]) -> Optional[tuple[int, ...]]:
                 sequence of integer tokens
         In case of empty tokens input data, None is returned
         """
-        pass
+        if not tokens:
+            return None
+
+        self._learn_indices(tokens)
+        numbers_tokens = []
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:155]: Do it in a single line

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_keywords_textrank/main.py (hunk 6 above)
Comment (d07l) [reply lab_3_keywords_textrank/main.py]: done

File: lab_3_keywords_textrank/main.py (hunk 5 above)
Comment (d07l) [reply lab_3_keywords_textrank/main.py]: fixed

File: lab_3_keywords_textrank/main.py [hunk 7]
```diff
@@ -136,7 +147,11 @@ def encode(self, tokens: tuple[str, ...]) -> Optional[tuple[int, ...]]:
                 sequence of integer tokens
         In case of empty tokens input data, None is returned
         """
-        pass
+        if not tokens:
+            return None
+        self._learn_indices(tokens)
+        numbers_tokens = [self._word2id[token] for token in tokens if token in self._word2id]
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:153]: generate tuple, not list

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_keywords_textrank/main.py (hunk 7 above)
Comment (d07l) [reply lab_3_keywords_textrank/main.py]: done

**Pull request closed without merging**